external/loaders/loaders/mappers/_hybrid.py

```diff
@@ -1,88 +1,190 @@
+    dt3dt_mp_coarse: xarray.DataArray
+    dt3dt_ogwd_coarse: xarray.DataArray
+    dt3dt_pbl_coarse: xarray.DataArray
+    dt3dt_shal_conv_coarse: xarray.DataArray
+    dt3dt_sw_coarse: xarray.DataArray
+    eddy_flux_vulcan_omega_sphum: xarray.DataArray
+    eddy_flux_vulcan_omega_temp: xarray.DataArray
+    exposed_area: xarray.DataArray
+    qv_dt_fv_sat_adj_coarse: xarray.DataArray
+    qv_dt_phys_coarse: xarray.DataArray
+    sphum: xarray.DataArray
+    sphum_storage: xarray.DataArray
+    sphum_vulcan_omega_coarse: xarray.DataArray
+    t_dt_fv_sat_adj_coarse: xarray.DataArray
+    t_dt_nudge_coarse: xarray.DataArray
+    t_dt_phys_coarse: xarray.DataArray
+    vulcan_omega_coarse: xarray.DataArray
+    T_vulcan_omega_coarse: xarray.DataArray
+
+
+def open_zarr(url, consolidated=False):
+    mapper = fsspec.get_mapper(url)
+    return xarray.open_zarr(mapper, consolidated=consolidated)
+
+
+def open_zarr_maybe_consolidated(url):
+    try:
+        return open_zarr(url, consolidated=True)
+    except KeyError:
+        return open_zarr(url, consolidated=False)
+
+
+def apparent_heating(data: FineResBudget):
+    eddy_flux = eddy_flux_coarse(
+        data.eddy_flux_vulcan_omega_temp,
+        data.T_vulcan_omega_coarse,
+        data.vulcan_omega_coarse,
+        data.T,
+    )
+    eddy_flux_convergence = convergence(eddy_flux, data.delp, dim="pfull")
+    return (
+        (data.t_dt_fv_sat_adj_coarse + data.t_dt_phys_coarse + eddy_flux_convergence)
+        .assign_attrs(
+            units="K/s",
+            long_name="apparent heating from high resolution data",
+            description=(
+                "Apparent heating due to physics and sub-grid-scale advection. Given "
+                "by "
+                "sat adjustment (dycore) + physics tendency + eddy-flux-convergence"
+            ),
         )
+        .rename("Q1")
+    )
 
-    def keys(self):
-        return list(
-            set(self.physics_mapper.keys()).intersection(self.fine_res_mapper.keys())
+
+def apparent_moistening(data: FineResBudget):
+    eddy_flux = eddy_flux_coarse(
+        data.eddy_flux_vulcan_omega_sphum,
+        data.sphum_vulcan_omega_coarse,
+        data.vulcan_omega_coarse,
+        data.sphum,
+    )
+    eddy_flux_convergence = convergence(eddy_flux, data.delp, dim="pfull")
+    return (
+        (data.qv_dt_fv_sat_adj_coarse + data.qv_dt_phys_coarse + eddy_flux_convergence)
+        .assign_attrs(
+            units="kg/kg/s",
+            long_name="apparent moistening from high resolution data",
+            description=(
+                "Apparent moistening due to physics and sub-grid-scale advection. "
+                "Given by "
+                "sat adjustment (dycore) + physics tendency + eddy-flux-convergence"
+            ),
         )
+        .rename("Q2")
+    )
 
-    def __len__(self):
-        return len(self.keys())
 
-    def __iter__(self):
-        return iter(self.keys())
+def _open_fine_res_dataset(
```

Comment:
I found this function name confusing given that it also opens a nudged run and takes some differences.

Reply:
Renamed to `open_fine_resolution_nudging_hybrid_dataset`.